File: agents/github-action-agent-node/src/hydrateGitMerge.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { execFileSync } from "node:child_process";
import {
  gitConflictBlocksToTonicAnnotatedPreview,
  parseGitConflicts,
  type GitMergeHydrationOptions,
} from "@mergetonic/core";
import type { HydratePair } from "./hydrate";

const BINARY = new Set([
  ".png",
  ".jpg",
  ".jpeg",
  ".gif",
  ".webp",
  ".ico",
  ".pdf",
  ".zip",
  ".gz",
  ".tgz",
  ".bz2",
  ".7z",
  ".rar",
  ".exe",
  ".dll",
  ".so",
  ".dylib",
  ".woff",
  ".woff2",
  ".ttf",
  ".eot",
  ".mp3",
  ".mp4",
  ".webm",
  ".wasm",
  ".pyc",
  ".class",
  ".jar",
]);

function isProbablyTextPath(filePath: string): boolean {
  const lower = filePath.toLowerCase();
  for (const ext of BINARY) {
    if (lower.endsWith(ext)) {
      return false;
    }
  }
  return true;
}

function gitOut(cwd: string, args: string[], allowFail = false): string {
  try {
    return execFileSync("git", args, { cwd, encoding: "utf8", stdio: ["ignore", "pipe", "pipe"] });
  } catch (error) {
    if (allowFail) {
      const e = error as { stdout?: string; stderr?: string };
      return [e.stdout ?? "", e.stderr ?? ""].join("\n");
    }
    throw error;
  }
}

function readStageLines(cwd: string, stage: "2" | "3", relPath: string): string[] {
  const out = gitOut(cwd, ["show", `:${stage}:${relPath}`], true);
  if (!out) {
    return [];
  }
  const lines = out.split(/\r?\n/);
  if (lines.length > 0 && lines[lines.length - 1] === "") {
    lines.pop();
  }
  return lines;
}

function hasConflictMarkers(lines: string[]): boolean {
  return lines.some((ln) => ln.startsWith("<<<<<<< "));
}

export type GitHydratePair = HydratePair & {
  gitAnnotatedLines?: string[];
  mergedLines?: string[];
};

export function hydrateGitMerge(params: {
  workspace: string;
  baseSha: string;
  headSha: string;
  maxFiles: number;
  gitMergeHydration?: GitMergeHydrationOptions;
}): Record<string, GitHydratePair> {
  const { workspace, baseSha, headSha, maxFiles, gitMergeHydration } = params;
  const isolatedWorkspace = (process.env.TONIC_AGENT_ISOLATED_WORKSPACE ?? "").trim();
  if (isolatedWorkspace && path.resolve(isolatedWorkspace) !== path.resolve(workspace)) {
    throw new Error("hydrateGitMerge workspace mismatch with TONIC_AGENT_ISOLATED_WORKSPACE");
  }
  if (!isolatedWorkspace && process.env.GITHUB_ACTIONS === "true") {
    throw new Error("hydrateGitMerge requires TONIC_AGENT_ISOLATED_WORKSPACE in GitHub Actions");
  }
  gitOut(workspace, ["checkout", "-f", baseSha]);
  const mergeOutput = gitOut(workspace, ["merge", "--no-ff", "--no-commit", headSha], true);
  const mergeFailed = /(^|\n)fatal:|(^|\n)error:/i.test(mergeOutput);
  const unmergedRaw = gitOut(workspace, ["diff", "--name-only", "--diff-filter", "U"], true);
  const paths = unmergedRaw
    .split(/\r?\n/)
    .map((s) => s.trim())
    .filter(Boolean);
  if (mergeFailed && paths.length === 0) {
    gitOut(workspace, ["merge", "--abort"], true);
    throw new Error(`git merge failed without unmerged files: ${mergeOutput.trim()}`);
  }
  const out: Record<string, GitHydratePair> = {};
  for (const relPath of paths) {
    if (Object.keys(out).length >= maxFiles) {
      break;
    }
    if (!isProbablyTextPath(relPath)) {
      continue;
    }
    const absPath = path.join(workspace, relPath);
    if (!fs.existsSync(absPath)) {
      continue;
    }
    const mergedText = fs.readFileSync(absPath, "utf8");
    const mergedLines = mergedText.split(/\r?\n/);
    if (mergedLines.length > 0 && mergedLines[mergedLines.length - 1] === "") {
      mergedLines.pop();
    }
    if (!hasConflictMarkers(mergedLines)) {
      continue;
    }
    const blocks = parseGitConflicts(mergedText);
    const annotated = gitConflictBlocksToTonicAnnotatedPreview(blocks, {
      repoRoot: workspace,
      leftRef: baseSha,
      rightRef: headSha,
      ...gitMergeHydration,
    });
    out[relPath] = {
      leftLines: readStageLines(workspace, "2", relPath),
      rightLines: readStageLines(workspace, "3", relPath),
      status: "unmerged",
      gitAnnotatedLines: annotated,
      mergedLines,
    };
  }
  gitOut(workspace, ["merge", "--abort"], true);
  return out;
}

File: agents/github-action-agent/src/tonic_agent/hydrate_git_merge.py
"""Hydrate files from a real git merge in an isolated workspace."""

from __future__ import annotations

import os
import subprocess
from pathlib import Path

from .git_merge_hydration import GitMergeHydrationOptions, git_blocks_to_tonic_annotated

BINARY_EXTENSIONS = frozenset(
    {
        ".png",
        ".jpg",
        ".jpeg",
        ".gif",
        ".webp",
        ".ico",
        ".pdf",
        ".zip",
        ".gz",
        ".tgz",
        ".bz2",
        ".7z",
        ".rar",
        ".exe",
        ".dll",
        ".so",
        ".dylib",
        ".woff",
        ".woff2",
        ".ttf",
        ".eot",
        ".mp3",
        ".mp4",
        ".webm",
        ".wasm",
        ".pyc",
        ".class",
        ".jar",
    }
)


def is_probably_text_path(path: str) -> bool:
    lower = path.lower()
    for ext in BINARY_EXTENSIONS:
        if lower.endswith(ext):
            return False
    return True


def _run_git(workspace: str, args: list[str], *, allow_fail: bool = False) -> str:
    cp = subprocess.run(
        ["git", *args],
        cwd=workspace,
        text=True,
        capture_output=True,
        check=False,
    )
    if cp.returncode != 0 and not allow_fail:
        raise RuntimeError(cp.stderr.strip() or cp.stdout.strip() or f"git {' '.join(args)} failed")
    output = (cp.stdout or "") + (("\n" + cp.stderr) if cp.stderr else "")
    return output


def _current_branch(workspace: str) -> str:
    out = _run_git(
        workspace,
        ["symbolic-ref", "--quiet", "--short", "HEAD"],
        allow_fail=True,
    ).strip()
    if not out or "fatal:" in out.lower():
        return ""
    return out.splitlines()[0].strip()


def _parse_git_conflicts(lines: list[str]) -> list[tuple[list[str], list[str]]]:
    blocks: list[tuple[list[str], list[str]]] = []
    i = 0
    while i < len(lines):
        if not lines[i].startswith("<<<<<<< "):
            i += 1
            continue
        i += 1
        left: list[str] = []
        right: list[str] = []
        while i < len(lines) and not lines[i].startswith("======="):
            left.append(lines[i])
            i += 1
        if i < len(lines):
            i += 1
        while i < len(lines) and not lines[i].startswith(">>>>>>> "):
            right.append(lines[i])
            i += 1
        if i < len(lines):
            i += 1
        blocks.append((left, right))
    return blocks


def _read_stage_lines(workspace: str, stage: str, rel_path: str) -> list[str]:
    out = _run_git(workspace, ["show", f":{stage}:{rel_path}"], allow_fail=True)
    if not out:
        return []
    lines = out.splitlines()
    if lines and lines[-1] == "":
        lines = lines[:-1]
    return lines


def hydrate_git_merge(
    *,
    workspace: str,
    base_sha: str,
    head_sha: str,
    max_files: int,
    git_merge_hydration: GitMergeHydrationOptions | None = None,
) -> dict[str, dict[str, object]]:
    isolated_workspace = (Path(str(workspace))).resolve()
    expected_iso = (
        Path(os.environ.get("TONIC_AGENT_ISOLATED_WORKSPACE", "")).resolve()
        if os.environ.get("TONIC_AGENT_ISOLATED_WORKSPACE")
        else None
    )
    if expected_iso is not None and isolated_workspace != expected_iso:
        raise RuntimeError(
            "hydrate_git_merge workspace mismatch with TONIC_AGENT_ISOLATED_WORKSPACE"
        )
    if expected_iso is None and os.environ.get("GITHUB_ACTIONS") == "true":
        raise RuntimeError("hydrate_git_merge requires TONIC_AGENT_ISOLATED_WORKSPACE in GitHub Actions")
    original_branch = _current_branch(workspace)
    try:
        _run_git(workspace, ["checkout", "-f", base_sha])
        merge_output = _run_git(workspace, ["merge", "--no-ff", "--no-commit", head_sha], allow_fail=True)
        names_raw = _run_git(workspace, ["diff", "--name-only", "--diff-filter", "U"], allow_fail=True)
        names = [n.strip() for n in names_raw.splitlines() if n.strip()]
        if ("fatal:" in merge_output.lower() or "error:" in merge_output.lower()) and not names:
            raise RuntimeError(f"git merge failed without unmerged files: {merge_output.strip()}")
        out: dict[str, dict[str, object]] = {}
        for rel in names:
            if len(out) >= max_files:
                break
            if not is_probably_text_path(rel):
                continue
            file_path = Path(workspace) / rel
            if not file_path.exists():
                continue
            merged_lines = file_path.read_text(encoding="utf-8", errors="replace").splitlines()
            if not any(ln.startswith("<<<<<<< ") for ln in merged_lines):
                continue
            blocks = _parse_git_conflicts(merged_lines)
            out[rel] = {
                "left_lines": _read_stage_lines(workspace, "2", rel),
                "right_lines": _read_stage_lines(workspace, "3", rel),
                "status": "unmerged",
                "annotated_lines": git_blocks_to_tonic_annotated(
                    blocks,
                    workspace=workspace,
                    base_sha=base_sha,
                    head_sha=head_sha,
                    opts=git_merge_hydration,
                ),
                "merged_lines": merged_lines,
            }
        return out
    finally:
        _run_git(workspace, ["merge", "--abort"], allow_fail=True)
        if original_branch:
            _run_git(workspace, ["checkout", "-f", original_branch], allow_fail=True)
